File: Leetcode/25. Reverse Nodes in k-Group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
private:
    void Reverse(ListNode* root) {
        if (root) {
            ListNode* curr = root, *prev = NULL, *next;
            while (curr) {
                next = curr -> next;
                curr -> next = prev;
                prev = curr;
                curr = next;
            }
        }
        return;
    }
    ListNode* helper(ListNode* head, int k) {
        if (head) {
            int size = k - 1;
            ListNode* root = head, *prev = head;
            while (head && size > 0) {
                size--;
                prev = head;
                head = head -> next;
            }
            if (head) {
                ListNode* nextNode = helper(head -> next, k);
                head -> next = NULL;
                Reverse(root);
                root -> next = nextNode;
                return head;
            }
            return root;
        }
        return NULL;
    }
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        return helper(head, k);
    }
};

File: Leetcode/25. Reverse Nodes in k-Group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        return self.helper(head, k)
        
    def Reverse(self, head: Optional[ListNode]):
        currNode, prevNode, nextNode = head, None, None
        while currNode:
            nextNode = currNode.next
            currNode.next = prevNode
            prevNode = currNode
            currNode = nextNode
        return
    
    def helper(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if head:
            size = k - 1
            root, prev = head, head
            while head and size > 0:
                size -= 1
                prev = head
                head = head.next
            if head:
                nextNode = self.helper(head.next, k)
                head.next = None
                self.Reverse(root)
                root.next = nextNode
                return head
            return root
        return None